pkg/addonmanager/controllers/clustermanagement/controller.go

```diff
@@ -0,0 +1,137 @@
+// based on the clustermanagementaddon.
+type clusterManagementController struct {
+	addonClient                  addonv1alpha1client.Interface
+	managedClusterLister         clusterlister.ManagedClusterLister
+	managedClusterAddonLister    addonlisterv1alpha1.ManagedClusterAddOnLister
+	clusterManagementAddonLister addonlisterv1alpha1.ClusterManagementAddOnLister
+	agentAddons                  map[string]agent.AgentAddon
+	eventRecorder                events.Recorder
+}
+
+func NewClusterManagementController(
+	addonClient addonv1alpha1client.Interface,
+	clusterInformers clusterinformers.ManagedClusterInformer,
+	addonInformers addoninformerv1alpha1.ManagedClusterAddOnInformer,
+	clusterManagementAddonInformers addoninformerv1alpha1.ClusterManagementAddOnInformer,
+	agentAddons map[string]agent.AgentAddon,
+	recorder events.Recorder,
+) factory.Controller {
+	c := &clusterManagementController{
+		addonClient:                  addonClient,
+		managedClusterLister:         clusterInformers.Lister(),
+		managedClusterAddonLister:    addonInformers.Lister(),
+		clusterManagementAddonLister: clusterManagementAddonInformers.Lister(),
+		agentAddons:                  agentAddons,
+		eventRecorder:                recorder.WithComponentSuffix(fmt.Sprintf("cluster-management-addon-controller")),
+	}
+
+	return factory.New().WithFilteredEventsInformersQueueKeyFunc(
+		func(obj runtime.Object) string {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			return key
+		},
+		func(obj interface{}) bool {
+			accessor, _ := meta.Accessor(obj)
+			if _, ok := c.agentAddons[accessor.GetName()]; !ok {
+				return false
+			}
+
+			return true
+		},
+		addonInformers.Informer(), clusterManagementAddonInformers.Informer()).
+		WithSync(c.sync).ToController(fmt.Sprintf("cluster-management-addon-controller"), recorder)
+}
+
+func (c *clusterManagementController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	key := syncCtx.QueueKey()
+	klog.V(4).Infof("Reconciling addon %q", key)
+
+	namespace, addonName, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		// ignore addon whose key is invalid
+		return nil
+	}
+
+	clusterManagementAddon, err := c.clusterManagementAddonLister.Get(addonName)
+	switch {
+	case errors.IsNotFound(err):
+		return nil
+	case err != nil:
+		return err
+	}
+
+	if len(namespace) == 0 {
+		return c.syncAllAddon(syncCtx, addonName)
+	}
+
+	coordinate := addonapiv1alpha1.ConfigCoordinates{
+		CRDName: clusterManagementAddon.Spec.AddOnConfiguration.CRDName,
+		CRName:  clusterManagementAddon.Spec.AddOnConfiguration.CRName,
+	}
+
+	addon, err := c.managedClusterAddonLister.ManagedClusterAddOns(namespace).Get(addonName)
+	switch {
+	case errors.IsNotFound(err):
+		return nil
+	case err != nil:
+		return err
+	}
+
+	addon = addon.DeepCopy()
+	if equality.Semantic.DeepEqual(coordinate, addon.Status.AddOnConfiguration) {
+		return nil
+	}
+
+	addon.Status.AddOnConfiguration = coordinate
```

Comment:
according to the api doc from https://github.com/open-cluster-management-io/api/blob/22c45bd6c7a17131978bb7b99543205734d7c1e3/addon/v1alpha1/types_managedclusteraddon.go#L90

the filed `status.RelatedObjects` is supposed to contain the information of a corresponding `ClusterManagementAddon` resource. how do you think about implement it together in this pull?